scheduler/src/cook/mesos/rebalancer.clj

```diff
@@ -391,64 +391,66 @@
                    jobs-to-make-room-for
-                   pending-job-ents-to-run
-                   task-ents-to-preempt)))
-
+                   preemption-decisions)))
         (do
           (timers/stop timer)
-          (histograms/update! task-counts-to-preempt (count task-ents-to-preempt))
-          (histograms/update! job-counts-to-run (count pending-job-ents-to-run))
-          ;; pending-job-ents-to-run is only for debugging purpose
-          [pending-job-ents-to-run task-ents-to-preempt]
-          )))))
+          (histograms/update! task-counts-to-preempt (count (mapcat :task preemption-decisions)))
+          (histograms/update! job-counts-to-run (count preemption-decisions))
+
+          preemption-decisions)))))
 
 (defn rebalance!
   [conn driver offer-cache pending-job-ents host->spare-resources params]
   (try
     (log/info "Rebalancing...Params:" params)
     (let [db (mt/db conn)
-          [pending-job-ents-to-run task-ents-to-preempt] (rebalance db offer-cache pending-job-ents host->spare-resources params)]
-      (log/info "Jobs to run:" pending-job-ents-to-run)
-      (log/info "Tasks to preempt:" task-ents-to-preempt)
-      (doseq [task-ent task-ents-to-preempt]
-        (try
-          @(d/transact
-            conn
-            ;; Make :instance/status and :instance/preempted? consistent to simplify the state machine.
-            ;; We don't want to deal with {:instance/status :instance.status/running, :instance/preempted? true}
-            ;; all over the place.
-            (let [job-eid (:db/id (:job/_instance task-ent))
-                  task-eid (:db/id task-ent)]
-              [[:generic/ensure task-eid :instance/status (d/entid db :instance.status/running)]
-               [:generic/atomic-inc job-eid :job/preemptions 1]
-               ;; The database can become inconsistent if we make multiple calls to :instance/update-state in a single
-               ;; transaction; see the comment in the definition of :instance/update-state for more details
-               [:instance/update-state task-eid :instance.status/failed [:reason/name :preempted-by-rebalancer]]
-               [:db/add task-eid :instance/reason [:reason/name :preempted-by-rebalancer]]
-               [:db/add task-eid :instance/preempted? true]]))
-          (catch Throwable e
-            (log/warn e "Failed to transact preemption")))
-        (when-let [task-id (:instance/task-id task-ent)]
-          (mesos/kill-task! driver {:value task-id}))))))
+          preemption-decisons (rebalance db offer-cache pending-job-ents host->spare-resources params)]
+      (doseq [{job-ent-to-make-room-for :to-make-room-for
+               task-ents-to-preempt :task} preemption-decisons]
+        ;; Ensure that uuids are loaded in entity
+        (:job/uuid job-ent-to-make-room-for)
+        (doall (map :instance/task-id task-ents-to-preempt))
+
+        (log/info "Preempting tasks to make room for waiting job"
```

Comment:
Here is an example:

```
2017-07-11 13:32:20,009 [async-thread-macro-2] INFO cook.mesos.rebalancer - Preempting tasks to make room for waiting job {:to-make-room-for {:db/id 17592186045652, :job/resour
ce #{{:db/id 17592186045653, :resource/type :resource.type/cpus, :resource/amount 2.0} {:db/id 17592186045654, :resource/type :resource.type/mem, :resource/amount 2000.0}}, :jo
b/priority 50, :job/user b, :job/uuid #uuid "59650b9a-f364-4a13-bb7c-96a6312a2719"}, :to-preempt ({:db/id 17592186045622, :job/_instance {:db/id 17592186045518, :job/resource #
{{:db/id 17592186045520, :resource/type :resource.type/mem, :resource/amount 2000.0} {:db/id 17592186045519, :resource/type :resource.type/cpus, :resource/amount 2.0}}, :job/us
er a, :job/priority 50}, :instance/start-time #inst "2017-07-11T17:32:10.971-00:00", :instance/hostname 156eed6c-fe44-4ed7-971e-e7b4432638b4, :instance/slave-id 156eed6c-fe44-4
ed7-971e-e7b4432638b4, :instance/task-id bfd30118-d811-4807-8e44-f44f3cbe0645})}
```

I went back and forth on this. On the one hand it would be a lot cleaner to reduce to the just the uuids, on the other hand, since this includes the resource info it makes it easier to understand what happened. I'm still largely on the fence so if you have a strong opinion I'm happy to switch it.